Lattice dynamics workflow using Pheasy

hrushikesh-s opened the pull request:
## Summary

Include a summary of major changes in bullet points:

* Feature 1 -- Add lattice dynamics workflow using pheasy for harmonic & anharmonic phonon calculations

## Additional dependencies introduced (if any)
* pheasy – used to generate symmetry-aware cluster spaces, construct null-space constraints, build displacement/sensing matrices, and fit harmonic force constants from displacement–force data; also used for optional higher-order IFC fitting, phonon renormalization, and force-constant conversion workflows.
* alamode (alm) – used to estimate the number of irreducible force-constant parameters and thereby determine how many displaced supercells are needed for harmonic and anharmonic IFC extraction; also used to estimate higher-order anharmonic parameter counts.
* hiphive – used as a post-processing and recovery backend for force constants: it enforces rotational sum rules on second-order IFCs when imaginary modes are present, reconstructs corrected force constants, and is also used to convert third-order force constants to phono3py format for thermal-conductivity workflows.

Before a pull request can be merged, the following items must be checked:

* [X] Code is in the [standard Python style](https://www.python.org/dev/peps/pep-0008/).
  The easiest way to handle this is to run the following in the **correct sequence** on
  your local machine. Start with running [`ruff`](https://docs.astral.sh/ruff) and `ruff format` on your new code. This will
  automatically reformat your code to PEP8 conventions and fix many linting issues.
* [X] Doc strings have been added in the [Numpy docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html).
  Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [X] Type annotations are **highly** encouraged. Run [mypy](http://mypy-lang.org) to
  type check your code.
* [X] Tests have been added for any new functionality or bug fixes.
* [X] All linting and tests pass.

leslie-zheng committed:
Add pheasy branch for phonon calculation using LASSO

leslie-zheng committed:
pheasy_phonons

leslie-zheng committed:
add_pheasy_to_vasp_folder

leslie-zheng committed:
Add_pheasy_to_vasp_file

JaGeo committed:
Merge branch 'main' into atomate2_jz_pheasy

JaGeo committed:
Merge branch 'main' into atomate2_jz_pheasy

leslie-zheng committed:
Modified some parts based on Janine's comments.

Modified some parts based on Janine's comments.

leslie-zheng committed:
Merge branch 'atomate2_jz_pheasy' of https://github.com/leslie-zheng/atomate2 into atomate2_jz_pheasy

leslie-zheng committed:
directly import the class methods from phonons

leslie-zheng committed:
clean up the code

leslie-zheng committed:
remove some files.

leslie-zheng committed:
minor change

leslie-zheng committed:
allow the users to define the number of displacements for random-displacement method.

leslie-zheng committed:
format_adjustment

leslie-zheng committed:
format_adjustment

leslie-zheng committed:
small update

leslie-zheng committed:
clean up the code and add more comments

leslie-zheng committed:
minor update

leslie-zheng committed:
allow the users to control the symmetry precision

leslie-zheng committed:
Lower the symmetry precision to allow pheasy to find a correct space group for crystals.

leslie-zheng committed:
Minor update

leslie-zheng committed:
minor adjustment

leslie-zheng committed:
minor update for flow/pheasy.py

leslie-zheng committed:
resuse some jobs from phonons

leslie-zheng committed:
clean up the job(generate_phonon_displacements)

leslie-zheng committed:
finished cleaning up the pheasy jobs module

leslie-zheng committed:
finished cleaning up the shemas/pheasy

leslie-zheng committed:
small adjustment to pass the lint

leslie-zheng committed:
raise a error if ALM is not installed.

leslie-zheng committed:
minor update

leslie-zheng committed:
Merge branch 'main' into atomate2_jz_pheasy

leslie-zheng committed:
first try

leslie-zheng committed:
finished testing

leslie-zheng committed:
Extract_higher_order_FCs

leslie-zheng committed:
anharmonic FCs cals. still in experiment.

leslie-zheng committed:
minor updata for the anharmonic FCs calculations

leslie-zheng committed:
minor update

leslie-zheng committed:
minor update

leslie-zheng committed:
minor update

leslie-zheng committed:
minor update

leslie-zheng committed:
adjustment and add comments

hrushikesh-s committed:
added the missing argument mp-id

hrushikesh-s committed:
added support of force field

leslie-zheng committed:
multiply a factor of 1.89 to convert the unit of A to Borh for ALM

If it did not convert, will generate a big mount of displacements for higher-order force constants.

leslie-zheng committed:
use filed to define fcs_cutoff_radius instead of a simple list.

it should work now @hrushikesh-s

leslie-zheng committed:
update definition of fcs_cutoff_radius

hrushikesh-s committed:
convert Bohr radius to Å

leslie-zheng committed:
giving a reasonable displaced distance for the anharmonic case.

leslie-zheng committed:
Merge branch 'atomate2_jz_pheasy_anharmonic' of https://github.com/leslie-zheng/atomate2 into atomate2_jz_pheasy_anharmonic

leslie-zheng committed:
minor update

leslie-zheng committed:
update

hrushikesh-s committed:
Unit tests for pheasy wf

hrushikesh-s committed:
Test # 1

1. updated the test
2. added the test data

leslie-zheng committed:
avoid circular import

leslie-zheng committed:
Merge branch 'atomate2_jz_pheasy_anharmonic' of https://github.com/leslie-zheng/atomate2 into atomate2_jz_pheasy_anharmonic

leslie-zheng committed:
update some stupid over-writing

leslie-zheng committed:
update the matrix I should use

leslie-zheng committed:
using phono3py for lattice thermal conductivity calculation

leslie-zheng committed:
update for phono3py thermal conductivity

leslie-zheng committed:
update phonon energy renomralization

leslie-zheng committed:
phonon renormalization

hrushikesh-s committed:
Auto stash before rebase of "atomate2_jz_pheasy_anharmonic" onto "origin/atomate2_jz_pheasy_anharmonic"

hrushikesh-s committed:
test files for pheasy

1. VASP input & output files
2. test_pheasy.py file defining the test

hrushikesh-s committed:
added 1 extra disp

hrushikesh-s committed:
ruff fixes to schemas/pheasy.py

hrushikesh-s committed:
ruff fixes to flows/pheasy.py

hrushikesh-s committed:
ruff fixes to jobs/pheasy.py

1. also modified the logic of assigning primitive_matrix

hrushikesh-s committed:
pre-commit vasp/flows/pheasy

hrushikesh-s committed:
updated pyproject.toml

hrushikesh-s committed:
Merge branch 'remote-main' into atomate2_jz_pheasy_anharmonic

hrushikesh-s committed:
updated the way to install pheasy for testing purposes

hrushikesh-s committed:
removed pheasy from [strict]

hrushikesh-s committed:
added hiphive dependency

hrushikesh-s committed:
added max_length argument

hrushikesh-s committed:
removed forcefields/test_pheasy.py

hrushikesh-s committed:
add conda install steps for alamode

JaGeo committed:
bug fix for linear strain passing in qha (#1061)

* bug fix for linear strain qha maker

* refactor to only rely on one method to generate supercells

* fix tests

* add test

* add test 2

* fix type annotation

* fix type annotation

* fix docs

* fix docs

* add test for supercell

* fix supercell tests

* fix supercell tests

hrushikesh-s committed:
updated conda command for alamode

hrushikesh-s committed:
adding f90nml to strict

hrushikesh-s committed:
adding further installation steps for alamode

hrushikesh-s committed:
alamode installation -- bug fix 1

hrushikesh-s committed:
alamode installation -- bug fix 2

hrushikesh-s committed:
alamode installation -- bug fix 3

hrushikesh-s commented:
Pheasy workflow depends on [Alamode](https://alamode.readthedocs.io/en/latest/install.html). @janosh, can you pls help in installing Alamode via `.github/workflows/testing.yml` file?

janosh commented:
no idea, never used or installed Alamode before. the best solution would be for them to publish pre-compiled wheels to PyPI  so the code doesn't need to be re-compiled from scratch on every CI run as that's both slow and tends to be brittle

JaGeo commented:
@naik-aakash set up a docker for https://github.com/autoatml/autoplex but it has been a massive amount of work. Before this, we did the compilation on the repo. The slowness was a real problem. But not sure it is realistic to get the wheels.

janosh commented:
Could offer an issue bounty to get buy in. Happy to pitch 50$

JaGeo commented:
@janosh thank you for offering this.

I am unsure about the timescales for such solutions. We would like to have the workflow available as soom as possible due to the planned atomate2 paper.

hrushikesh-s commented:
In that case, @JaGeo & @naik-aakash can you pls help with the compilation of alamode on the repo?
Also tagging in @leslie-zheng.
> @naik-aakash set up a docker for https://github.com/autoatml/autoplex but it has been a massive amount of work. Before this, we did the compilation on the repo. The slowness was a real problem. But not sure it is realistic to get the wheels. To sum up: compilation on the repo might be a realistic and okay solution if it isn't too slow.

hrushikesh-s commented:
These are the exact commands to install Alamode:

```
1. conda install -c conda-forge numpy scipy h5py compilers “libblas=*=*mkl” spglib boost eigen cmake ipython mkl-include
2. git clone https://github.com/ttadano/ALM.git
3. cd ALM
4. python setup.py build
5. pip install -e .
```

JaGeo commented:
Just add it to the workflow file. You can see how micromamba is used for installation here:
https://github.com/materialsproject/atomate2/blob/3d55b2b56b2dfb938ab60b6deb198b8e682fcd7f/.github/workflows/testing.yml#L61

hrushikesh-s committed:
alamode installation -- bug fix 4

JaGeo commented on src/atomate2/common/flows/phonons.py:
```diff
@@ -162,6 +162,7 @@ class BasePhononMaker(Maker, ABC):
 
     kpath_scheme: str = "seekpath"
     code: str = None
+    mp_id: str = None
```
This can be added as metadata.

hrushikesh-s replied:
Done!

hrushikesh-s committed:
alamode installation -- bug fix 5

hrushikesh-s committed:
alamode installation -- bug fix 6

janosh committed:
Fix `generate_phonon_displacements`  magmom removal side effect (#1064)

* fix lints and typos

* fix generate_phonon_displacements side effect: removes magmoms of passed structure in-place

* fix ruff PYI063 Use PEP 570 syntax for positional-only parameters

hrushikesh-s committed:
added tol to test_phonons.py

JaGeo commented on tests/vasp/flows/test_phonons.py:
```diff
@@ -202,6 +202,8 @@ def test_phonon_wf_vasp_only_displacements_no_structural_transformation(
     assert_allclose(
         responses[job.jobs[-1].uuid][1].output.primitive_matrix,
         ((0, 1, 0), (0, 0, 1), (1, 0, 0)),
+        rtol=1e-8,
```
@hrushikesh-s @naik-aakash and I are working on fixing these tests. I think I messed something up with the test files. I might have time later tonight. So, maybe skip those tests first.

JaGeo replied:
Aah. Sorry! Different test 😅. But I meant the update to the latest phonopy. Sorry!

hrushikesh-s replied:
oh okay okay. I was wondering why suddenly the test_phonons.py file was failing :sweat_smile:

dependabot[bot] committed:
Bump emmet-core from 0.84.3rc3 to 0.84.3rc4 (#1065)

Bumps [emmet-core](https://github.com/materialsproject/emmet) from 0.84.3rc3 to 0.84.3rc4.
- [Release notes](https://github.com/materialsproject/emmet/releases)
- [Changelog](https://github.com/materialsproject/emmet/blob/main/docs/CHANGELOG.md)
- [Commits](https://github.com/materialsproject/emmet/compare/0.84.3rc3...0.84.3rc4)

---
updated-dependencies:
- dependency-name: emmet-core
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump jobflow from 0.1.18 to 0.1.19 (#1067)

Bumps [jobflow](https://github.com/materialsproject/jobflow) from 0.1.18 to 0.1.19.
- [Release notes](https://github.com/materialsproject/jobflow/releases)
- [Changelog](https://github.com/materialsproject/jobflow/blob/main/CHANGELOG.md)
- [Commits](https://github.com/materialsproject/jobflow/compare/v0.1.18...v0.1.19)

---
updated-dependencies:
- dependency-name: jobflow
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

JaGeo committed:
add openmm description to docs (#1069)

janosh committed:
fix minor code formatting issues + typos (#1070)

esoteric-ephemera committed:
Docs update, forcefield elastic convenience maker, forcefield enum hydration (#1072)

* update EOS docs

* udpate docs with implementation details

* add convenience constructor method for forcefield ElasticMaker

* allow MLFF enum to treat str(MLFF.) as valid member

* precommit

* add small MLFF test

naik-aakash committed:
Fix bump phonopy (#1006)

* bump phonopy version

* Update pyproject.toml

* replace bs plotter with pymatgen implementation > reduce code repetition

* replace bs plotter with pymatgen implementation > reduce code repetition

* switch off grunesien computation at gamma (unstable)

* fix assertions to correct expected derived properties

* fix failing test (use correct reference data)

* remove non used test_data, rename Si_phonons_4 > Si_phonons_3

* point to renamed ref test_data

---------

Co-authored-by: J. George <JaGeo@users.noreply.github.com>

dependabot[bot] committed:
Bump mp-api from 0.42.2 to 0.43.0 (#1075)

Bumps mp-api from 0.42.2 to 0.43.0.

---
updated-dependencies:
- dependency-name: mp-api
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump pytest from 8.3.3 to 8.3.4 (#1077)

Bumps [pytest](https://github.com/pytest-dev/pytest) from 8.3.3 to 8.3.4.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/8.3.3...8.3.4)

---
updated-dependencies:
- dependency-name: pytest
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump ipython from 8.29.0 to 8.30.0 (#1076)

Bumps [ipython](https://github.com/ipython/ipython) from 8.29.0 to 8.30.0.
- [Release notes](https://github.com/ipython/ipython/releases)
- [Commits](https://github.com/ipython/ipython/compare/8.29.0...8.30.0)

---
updated-dependencies:
- dependency-name: ipython
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump sevenn from 0.10.1 to 0.10.2 (#1080)

Bumps [sevenn](https://github.com/MDIL-SNU/SevenNet) from 0.10.1 to 0.10.2.
- [Release notes](https://github.com/MDIL-SNU/SevenNet/releases)
- [Changelog](https://github.com/MDIL-SNU/SevenNet/blob/main/CHANGELOG.md)
- [Commits](https://github.com/MDIL-SNU/SevenNet/compare/v0.10.1...v0.10.2)

---
updated-dependencies:
- dependency-name: sevenn
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump emmet-core from 0.84.3rc4 to 0.84.3rc6 (#1081)

Bumps [emmet-core](https://github.com/materialsproject/emmet) from 0.84.3rc4 to 0.84.3rc6.
- [Release notes](https://github.com/materialsproject/emmet/releases)
- [Changelog](https://github.com/materialsproject/emmet/blob/main/docs/CHANGELOG.md)
- [Commits](https://github.com/materialsproject/emmet/compare/0.84.3rc4...v0.84.3rc6)

---
updated-dependencies:
- dependency-name: emmet-core
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

naik-aakash committed:
update calorine version (#1083)

hrushikesh-s committed:
soften the assert output["relax"]["EOS"]["birch_murnaghan"]["b0"]

hrushikesh-s committed:
further softening

hrushikesh-s committed:
providing absolute path for FC file

hrushikesh-s committed:
changed the location of reading FC file

hrushikesh-s committed:
changed the run_locally's attribute of create_folders to false

hrushikesh-s committed:
changed it back to True

hrushikesh-s committed:
adding some print statements for debugging

hrushikesh-s committed:
adding more logger statements for debugging

hrushikesh-s commented:
@JaGeo , The test fails because GH can't find FORCECONSTANTS file in the temp folder. However, when I run the same test locally, I am able to successfully run it. Can you pls help?

JaGeo commented:
@hrushikesh-s I will try to look into it. Could you add all installation instructions to the documentation as well?

JaGeo commented on .github/workflows/testing.yml:
```diff
@@ -67,8 +67,20 @@ jobs:
+          micromamba activate a2
+          micromamba install -n a2 -c conda-forge numpy scipy h5py compilers “libblas=*=*mkl” spglib boost eigen cmake ipython mkl-include openmpi --yes  
+          git clone https://github.com/ttadano/ALM.git # do I need to modify this?
+          cd ALM # do I need to modify this?
+          cd python
```
don't you need the develop branch? I currently cannot install it locally

JaGeo commented:
@hrushikesh-s Locally, it fails for me with exactly the same error. Are you sure you have the same pheasy, ALM and phonopy versions? I only have a force_matrix.pkl file in the last folder but no FORCE_CONSTANT file

JaGeo commented:
I also got a numpy error. Something in the pheasy commands seems to go wrong.

hrushikesh-s commented:
> @hrushikesh-s I will try to look into it. Could you add all installation instructions to the documentation as well?
Yes. 

Doing this now

> @hrushikesh-s Locally, it fails for me with exactly the same error. Are you sure you have the same pheasy, ALM and phonopy versions? I only have a force_matrix.pkl file in the last folder but no FORCE_CONSTANT file

Checking this now

hrushikesh-s committed:
commenting out the anh portion of the code that is buggy

replacing it with a constant num_dis_cells_anhar = 20

hrushikesh-s committed:
updating get_supercell_size()

1. added this function into jobs/pheasy.py such that now max_length is no longer compulsorily required.
2. the default values of min_length is changed to 8 & max_atoms to 200

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -31,6 +34,40 @@
 
 logger = logging.getLogger(__name__)
 
+@job
```
What is the key difference here? Happy to modify the normal supercell search as well if there are weaknesses

hrushikesh-s replied:
I wanted the get_supercell_size() to accept only min_length, max_atoms, force_90_degrees, force_diagonal, and allow_orthorhombic, but it currently also insists on a max_length argument.

That's why, as a quick fix, I created this new function. 😅

But if the original get_supercell_size() from jobs/phonons.py can be modified such that it doesn't insists on max_length argument, then that'll also work 😅

PS: For my current HT phonon runs (nsites ≥ 41) I specifically wanted:
min_length = 8.0
max_atoms = 200
force_90_degrees = True
force_diagonal = True
allow_orthorhombic = False

Tagging @leslie-zheng as well

JaGeo replied:
I see. Thanks. Yes, maybe we should make the modification once this workflow is completely implemented and tested!

JaGeo replied:
can we merge this with the "get_supercell_size"? I think if it is just about the max_length, it must be possible to get this modified.

hrushikesh-s committed:
The file name is "FORCE_CONSTANTS"

hrushikesh-s committed:
added variables required for the get_supercell job

JaGeo commented:
@hrushikesh-s Thank you for this pull-request.
I have a list of TODOs:

- [ ] Solve conflicts with the current main branch
- [ ] Check if the tests run through here on github actions 
- [ ] Add a real test for pheasy (only see schema tests so far)
- [ ] Add new documentation

Let me know if you need help with individual parts.

leslie-zheng commented:
> @hrushikesh-s Thank you for this pull-request. I have a list of TODOs:
> 
> * [ ]  Solve conflicts with the current main branch
> * [ ]  Check if the tests run through here on github actions
> * [ ]  Add a real test for pheasy (only see schema tests so far)
> * [ ]  Add new documentation
> 
> Let me know if you need help with individual parts.

Hi Janine, Could you please tell us more about the requirement of new document? @JaGeo 
thanks 
jiongzhi

JaGeo commented:
@leslie-zheng you mean documentation? Just some info on the approach as a part of the docs. Happy to help you write it once the conflicts are solved and a new test is there.

hrushikesh-s committed:
added the preserve_fworker tag

This ensure that all the children jobs are run on the same HPC

hrushikesh-s committed:
added the support for hdf5 file

1. hdf5 file format is faster to load as compared to yaml file format. hence, writing the band structure in this format as well.

JaGeo commented on src/atomate2/common/flows/pheasy.py:
```diff
@@ -0,0 +1,446 @@
+            Instead of min_length, also a supercell_matrix can be given, e.g.
+            [[1.0,0.0,0.0],[0.0,1.0,0.0],[0.0,0.0,1.0]
+        """
+        use_symmetrized_structure = self.use_symmetrized_structure
+        kpath_scheme = self.kpath_scheme
+        valid_structs = (None, "primitive", "conventional")
+        if use_symmetrized_structure not in valid_structs:
+            raise ValueError(
+                f"Invalid {use_symmetrized_structure=}, use one of {valid_structs}"
+            )
```
There is some code repetition here from the harmonic PhononMaker. I think we should either inherit from this Maker or move those code blocks into functions that can be easily reused.

hrushikesh-s replied:
Okay, makes sense. How should we do it? Should these code blocks be moved into a new function in common/jobs/phonons.py ?

JaGeo commented on src/atomate2/common/flows/pheasy.py:
```diff
@@ -0,0 +1,446 @@
+            jobs.append(compute_total_energy_job)
+            total_dft_energy = compute_total_energy_job.output
+
+        # get a phonon object from pheasy code using the random-displacement approach
+        displacements = generate_phonon_displacements(
+            structure=structure,
+            supercell_matrix=supercell_matrix,
+            displacement=self.displacement,
+            num_displaced_supercells=self.num_displaced_supercells,
+            cal_anhar_fcs=self.cal_anhar_fcs,
+            displacement_anhar=self.displacement_anhar,
+            num_disp_anhar=self.num_disp_anhar,
+            fcs_cutoff_radius=self.fcs_cutoff_radius,
+            sym_reduce=self.sym_reduce,
+            symprec=self.symprec,
+            use_symmetrized_structure=self.use_symmetrized_structure,
+            kpath_scheme=self.kpath_scheme,
+            code=self.code,
+        )
```
This part is now different than in the harmonic phonon maker. Maybe, one could add a method to the phonon.BasePhononMaker and overwriting this here with the new implementation.

hrushikesh-s replied:
Didnt understand this. Do you mean writing a method in common/flows/phonons.py ?

JaGeo commented on src/atomate2/common/flows/pheasy.py:
```diff
@@ -0,0 +1,446 @@
+        jobs.append(displacements)
+
+        # perform the phonon displacement calculations
+        displacement_calcs = run_phonon_displacements(
+            displacements=displacements.output,
+            structure=structure,
+            supercell_matrix=supercell_matrix,
+            phonon_maker=self.phonon_displacement_maker,
+            socket=self.socket,
+            prev_dir_argname=self.prev_calc_dir_argname,
+            prev_dir=prev_dir,
+        )
```
This is also different than in phonon.BasePhononMaker

hrushikesh-s replied:
This is currently getting imported like this:
```
from atomate2.common.jobs.pheasy import (
    generate_frequencies_eigenvectors,
    generate_phonon_displacements,
    get_supercell_size,
    run_phonon_displacements,
)
```

> This is also different than in phonon.BasePhononMaker

What should be done to fix this then?

JaGeo replied:
I will draft something 😃

hrushikesh-s committed:
Merge remote-tracking branch 'materialsproject/main' into atomate2_jz_pheasy_anharmonic

hrushikesh-s committed:
ruff fixes

hrushikesh-s committed:
added pheasy dependency to the GH actions

hrushikesh-s committed:
removed numpy from ALM installation on GH action

hrushikesh-s committed:
Need numpy<2.2 for hiphive

HipHive leverages Numba under the hood for its force-constant calculators, and numba only works with NumPy ≤ 2.2.

hrushikesh-s committed:
hiphive needs numpy < 2.2

hrushikesh-s committed:
changed the sequence of installing ALM

hrushikesh-s committed:
removed kwargs!

hrushikesh-s committed:
updated the way pheasy is installed in GH actions

hrushikesh-s committed:
updated pyproject.toml with Pheasy

hrushikesh-s committed:
removed the thermal displacement check from test_pheasy

JaGeo commented on src/atomate2/common/flows/pheasy.py:
```diff
@@ -0,0 +1,446 @@
+            born_run_job_dir = born_job.output.dir_name
+            born_run_uuid = born_job.output.uuid
+
+        phonon_collect = generate_frequencies_eigenvectors(
+            supercell_matrix=supercell_matrix,
+            displacement=self.displacement,
+            num_displaced_supercells=self.num_displaced_supercells,
+            cal_anhar_fcs=self.cal_anhar_fcs,
+            displacement_anhar=self.displacement_anhar,
+            num_disp_anhar=self.num_disp_anhar,
+            fcs_cutoff_radius=self.fcs_cutoff_radius,
+            renorm_phonon=self.renorm_phonon,
+            renorm_temp=self.renorm_temp,
+            cal_ther_cond=self.cal_ther_cond,
+            ther_cond_mesh=self.ther_cond_mesh,
+            ther_cond_temp=self.ther_cond_temp,
```
This would be one option but it might be a breaking change.

hrushikesh-s committed:
removed the mp-id from common/flows/phonons

JaGeo commented on .github/workflows/testing.yml:
```diff
@@ -60,16 +60,28 @@ jobs:
         run: |
           micromamba install -n a2 -c conda-forge enumlib packmol bader --yes
 
+      - name: Install ALA-Mode
+        run: |
+          micromamba activate a2
+          micromamba install -n a2 -c conda-forge "numpy<=2.2" scipy h5py compilers “libblas=*=*mkl” spglib boost eigen cmake ipython mkl-include openmpi --yes  
+          git clone https://github.com/ttadano/ALM.git # do I need to modify this?
+          cd ALM # do I need to modify this?
+          cd python
+          python setup.py build # do I need to modify this?
+          uv pip install -e . # do I need to modify this?
+
```
Could you add these instructions to the documentation?

JaGeo commented on .github/workflows/testing.yml:
```diff
@@ -60,16 +60,28 @@ jobs:
+        run: |
+          micromamba activate a2
+          micromamba install -n a2 -c conda-forge "numpy<=2.2" scipy h5py compilers “libblas=*=*mkl” spglib boost eigen cmake ipython mkl-include openmpi --yes  
+          git clone https://github.com/ttadano/ALM.git # do I need to modify this?
```
https://anaconda.org/conda-forge/alm why not installing it directly from conda? Is a different version needed?

JaGeo commented on src/atomate2/vasp/flows/pheasy.py:
```diff
@@ -0,0 +1,162 @@
+
+@dataclass
+class PhononMaker(BasePhononMaker):
+    """
+    Maker to calculate harmonic phonons with VASP and Phonopy.
+
+    Calculate the harmonic phonons of a material. Initially, a tight structural
+    relaxation is performed to obtain a structure without forces on the atoms.
+    Subsequently, supercells with one displaced atom are generated and accurate
+    forces are computed for these structures. With the help of phonopy, these
+    forces are then converted into a dynamical matrix. To correct for polarization
+    effects, a correction of the dynamical matrix based on BORN charges can
+    be performed. Finally, phonon densities of states, phonon band structures
+    and thermodynamic properties are computed.
```
Doc strings would need to be updated.

JaGeo committed:
restructure

JaGeo committed:
fix doc strings

JaGeo committed:
Merge branch 'main' into atomate2_jz_pheasy_anharmonic

JaGeo committed:
fix doc strings

JaGeo committed:
Merge branch 'atomate2_jz_pheasy_anharmonic' of github.com:leslie-zheng/atomate2 into atomate2_jz_pheasy_anharmonic

JaGeo committed:
fix mpid

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,475 @@
+        code to perform the computations
+
+    """
+    warnings.warn(
+        "Initial magnetic moments will not be considered for the determination "
+        "of the symmetry of the structure and thus will be removed now.",
+        stacklevel=1,
+    )
+    cell = get_phonopy_structure(
+        structure.remove_site_property(property_name="magmom")
```
ideally, we export the common parts from the phonons and pheasy jobs into a method and reuse it? This would then also make it easier when we change anything here in the future in either of the jobs in pheasy.py or phonons.py

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,475 @@
+    )
+
+
+# I did not directly import this job from the phonon module
+# because I modified the job to pass the displaced structures
+# to the output.
+@job(data=["forces", "displaced_structures"])
+def run_phonon_displacements(
+    displacements: list[Structure],
+    structure: Structure,
+    supercell_matrix: Matrix3D,
+    phonon_maker: BaseVaspMaker | ForceFieldStaticMaker | BaseAimsMaker = None,
+    prev_dir: str | Path = None,
+    prev_dir_argname: str = None,
+    socket: bool = False,
+) -> Flow:
```
if possible, i would modify the phonons.py job instead to avoid code duplication

JaGeo committed:
fix mpid

JaGeo committed:
fix mpid

JaGeo commented:
- [ ] We need to add the Pheasy reference to the documentation or the doc string of this PR: https://arxiv.org/abs/2508.01020

esoteric-ephemera committed:
sync with main

esoteric-ephemera committed:
fix supercell gen?

esoteric-ephemera committed:
remove builder meta

esoteric-ephemera committed:
update docs with basic pheasy info

esoteric-ephemera committed:
alamode install instructions

esoteric-ephemera committed:
clean up pheasy common jobs

esoteric-ephemera committed:
reduce code duplication between phonopy only and +pheasy jobs

esoteric-ephemera committed:
migrate to emmet core dependence

esoteric-ephemera committed:
clean up

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -189,61 +177,41 @@ def generate_phonon_displacements(
 
-    # Here, the ALM module is used to determine how many free parameters of third and
-    # fourth order force constants (FCs) within the specific supercell.
+    # Here, the ALAMODE copde is used to determine the number of
```
```suggestion
    # Here, the ALAMODE code is used to determine the number of
```

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -341,133 +306,608 @@ def generate_frequencies_eigenvectors(
 
-    Note, this job will replace itself with N displacement calculations,
-    or a single socket calculation for all displacements.
+    # TODO: potentially add kwargs to avoid computation of eigenvectors
```
```suggestion
```

esoteric-ephemera committed:
merge conflict

esoteric-ephemera committed:
tweak tests

esoteric-ephemera committed:
fix notebooks with emmet transition + qha analysis

esoteric-ephemera committed:
patch up phonon tests from schema migration

esoteric-ephemera committed:
add check / warning for openff install

esoteric-ephemera committed:
remove schema tests which now live in emmet-core; fix gruneisen

esoteric-ephemera committed:
Merge remote-tracking branch 'upstream/main' into atomate2_jz_pheasy_anharmonic

esoteric-ephemera committed:
ensure forcefield phonon test data is more reliable

esoteric-ephemera committed:
add glue job + schemas for qha

esoteric-ephemera committed:
install pheasy for docs

esoteric-ephemera committed:
the mystery of python tuples

esoteric-ephemera committed:
Merge branch 'main' into atomate2_jz_pheasy_anharmonic

esoteric-ephemera committed:
Merge remote-tracking branch 'upstream/main' into atomate2_jz_pheasy_anharmonic

esoteric-ephemera committed:
make default file list

esoteric-ephemera committed:
missed a few files

esoteric-ephemera committed:
funkiness in cwd

esoteric-ephemera committed:
merge upstream

esoteric-ephemera committed:
try debugging wf tests

esoteric-ephemera committed:
add file existence checks

hrushikesh-s committed:
Merge remote-tracking branch 'upstream/main' into atomate2_jz_pheasy_anharmonic

hrushikesh-s committed:
Enable matrix-based displacement and force input in pheasy job commands

Update pheasy job execution commands to explicitly pass displacement and force matrix files generated by atomate2.

Changes:
- Pass --disp_matrix_file and --force_matrix_file to pheasy during sensing matrix construction and IFC fitting
- Replace implicit file-based displacement/force reconstruction with
  explicit matrix-based input when available

hrushikesh-s committed:
Fix CI env setup and dependency installation for testing workflow

Update GitHub Actions workflow & pyproject.toml to fix env setup.

Changes:
- Consolidate micromamba dependency installation into a single command
- Add missing core dependencies for alamode (compilers, boost, eigen, cmake, spglib, scipy, h5py)
- Remove separate ALA-mode installation step
- Include alamode in pip installation dependencies

hrushikesh-s committed:
Remove duplicate PhononBSDOSDoc schema in favor of emmet import

Remove local definition of PhononBSDOSDoc from phonons.py.

This schema is already provided by emmet and imported in the beginning,
so maintaining a duplicate definition is unnecessary and can lead
to inconsistencies.

hrushikesh-s committed:
Add torchsim support to phonon workflows and stabilize forcefield phonon test

Changes:
- Added "torchsim" to the supported code list in `get_factor` so that phonon
  frequency conversion works correctly for TorchSim-based calculations.
- Updated `test_phonon_wf_force_field` to use `assert_allclose` with a relaxed
  tolerance instead of element-wise `pytest.approx`, reducing brittleness due
  to small numerical differences in thermodynamic quantities.

Rationale:
- TorchSim behaves similarly to forcefield-based calculators but was not
  previously recognized in the phonon conversion logic, leading to failures.
- The phonon's thermodynamic test was overly sensitive to small numerical
  variations (eg due to backend changes like CHGNet to matgl transition?), causing
  unnecessary test failures.

hrushikesh-s committed:
Fix pytest invocation in testing.yaml file

This commit updates the GitHub Actions test workflow to invoke pytest using
`python -m pytest` via `micromamba run`, instead of relying on shell activation
and the `pytest` executable.

Changes:
- Replaced `micromamba activate a2` + `pytest ...` with
  `micromamba run -n a2 python -m pytest ...` across all test steps.

Reason:
- Fixes CI failures where `pytest` was not found due to PATH issues after
  environment activation.
- Ensures tests are executed within the correct environment regardless of
  shell behavior.

hrushikesh-s committed:
removing the commented lines

hrushikesh-s committed:
added missing pytest dependency

hrushikesh-s committed:
revert back to original format as used in "main"

hrushikesh-s committed:
Use pheasy fork with relaxed ASE dependency for CI compatibility

hrushikesh-s committed:
Re-run CI with updated pheasy fork

hrushikesh-s committed:
Fix pheasy phonon tests and stabilize forcefield CI dependencies

- Update thermo reference values for matgl-backed phonon workflows
- Enable thermal displacement generation in pheasy test (create_thermal_displacements=True)
- Relax force_constants type check (tuple --> list | tuple)
- Update total_dft_energy expectation (per-atom normalization)
- Fix CI dependency stack for torch-limited forcefields

hrushikesh-s committed:
lint fix

hrushikesh-s committed:
lint fix

hrushikesh-s committed:
Restore legacy atomate2 phonon schemas

This commit restores support for the legacy atomate2 phonon document schemas while introducing optional support for the newer Emmet schemas across phonon-related workflows.

Deprecation decision:
No deprecation is introduced in this commit
Both schemas are supported going forward
Deprecation of legacy atomate2 phonon schemas is deferred and will be discussed separately with a proper timeline

hrushikesh-s committed:
fixes to make the phonon related workflows compatible with original atomate2 data schema

hrushikesh-s committed:
lint fix and fix to phonon aims tutorial

hrushikesh-s committed:
Backward compatibility for older atomate2 tests/callers

hrushikesh-s committed:
Merge branch 'main' into atomate2_jz_pheasy_anharmonic

JaGeo commented on src/atomate2/common/jobs/gruneisen.py:
```diff
@@ -50,7 +53,7 @@ def shrink_expand_structure(structure: Structure, perc_vol: float) -> Response:
     return Response(output={"plus": plus_struct, "minus": minus_struct})
 
 
-@job(data=[PhononBSDOSDoc])
```
With this new workflow, no object needs to be transferred to "data"?

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,953 @@
+        )
+
+    # Using a shorter cutoff (10 A) to generate the force constants to
+    # eliminate the imaginary modes near Gamma point in phesay code
```
```suggestion
    # eliminate the imaginary modes near Gamma point in pheasy code
```

JaGeo committed:
Apply suggestion from @JaGeo

JaGeo commented on src/atomate2/common/jobs/qha.py:
```diff
@@ -56,7 +60,7 @@ def get_supercell_size(
     )
 
 
-@job(data=[PhononBSDOSDoc])
```
Is this not needed?

JaGeo replied:
Or at least a replacement for this one?

JaGeo commented on tests/common/schemas/test_phonons.py:
```diff
@@ -1,60 +0,0 @@
-from monty.json import MontyEncoder
-from pydantic import ValidationError
-
-from atomate2.common.schemas.phonons import (
-    PhononBSDOSDoc,
-    PhononComputationalSettings,
-    PhononJobDirs,
-    PhononUUIDs,
-    ThermalDisplacementData,
-)
```
I understood that we keep this for now, yes? Why not keeping the tests as long as we keep the schema?

JaGeo commented on src/atomate2/common/flows/pheasy.py:
```diff
@@ -0,0 +1,362 @@
+"""Flow for calculating (an)harmonic FCs and phonon energy renorma. with pheasy."""
```
```suggestion
"""Flow for calculating (an)harmonic FCs and phonon energy renormalisation with pheasy."""
```

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,953 @@
+    if cal_anhar_fcs:
+        # Due to the cutoff radius of the force constants use the unit of Borh in ALM,
+        # we need to convert the cutoff radius from Angstrom to Bohr.
+        with ALM(lattice * 1.89, positions, numbers) as alm:
```
I assume this is accurate enough for a cutoff but maybe more digits?

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,953 @@
+        subprocess.call(shlex.split(pheasy_cmd_13))
+        subprocess.call(shlex.split(pheasy_cmd_14))
+
+        force_constants = parse_FORCE_CONSTANTS(filename=new_fc_file)
+        phonon.force_constants = force_constants
+        phonon.symmetrize_force_constants()
+
+        phonon.save(_DEFAULT_FILE_PATHS["phonopy"])
+
+        # get phonon band structure
+        kpath_dict, kpath_concrete = _get_kpath(
+            structure=get_pmg_structure(phonon.primitive),
+            kpath_scheme=kpath_scheme,
+            symprec=symprec,
+        )
+
+        npoints_band = kwargs.get("npoints_band", 101)
+        qpoints, connections = get_band_qpoints_and_path_connections(
+            kpath_concrete, npoints=kwargs.get("npoints_band", 101)
+        )
+
+        # phonon band structures will always be computed
+        phonon.run_band_structure(
+            qpoints, path_connections=connections, with_eigenvectors=True
+        )
+        phonon.write_yaml_band_structure(filename=_DEFAULT_FILE_PATHS["band_structure"])
+        bs_symm_line = get_ph_bs_symm_line(
+            _DEFAULT_FILE_PATHS["band_structure"],
+            labels_dict=kpath_dict,
+            has_nac=born is not None,
+        )
```
can more code be reused from phonon.py? If we have an update for the plotting, we would need to fix it here as well. Same for the code below.

JaGeo commented on src/atomate2/openff/__init__.py:
```diff
@@ -1,5 +1,13 @@
 """Module for classical md workflows."""
 
+from importlib.util import find_spec
+
+if not find_spec("openff"):
+    raise ImportError(
+        "openff must be installed via conda forge to use with atomate2:\n"
+        "conda install -c conda-force openff"
+    )
+
```
can you check if this is consistent with how we handle this accross atomate2?

JaGeo commented on tests/vasp/flows/test_pheasy.py:
```diff
@@ -0,0 +1,130 @@
+        ph_doc.supercell_matrix,
+        [[4.0, 0.0, 0.0], [0.0, 4.0, 0.0], [0.0, 0.0, 4.0]],
+    )
+    assert_allclose(
+        ph_doc.primitive_matrix,
+        [[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]],
+        rtol=1e-5,
+        atol=1e-10,
+    )
+    assert ph_doc.code == "vasp"
+    assert isinstance(
+        ph_doc.post_process_settings,
+        PhononComputationalSettings,
+    )
+    assert ph_doc.post_process_settings.npoints_band == 101
+    assert ph_doc.post_process_settings.kpath_scheme == "seekpath"
+    assert ph_doc.post_process_settings.kpoint_density_dos == 7000
+
+    assert ph_doc.chemsys == "Si"
```
I don't see a test that actually checks the new implemented functionality. How do we make sure that the results stay consistent?

JaGeo commented:
Hi  @hrushikesh-s ,

I left a few comments.

My main points:

- [ ] pheasy functionality must be tested, e.g., by testing some of the phonon results.
- [ ] could you check that both Grüneisen and QHA workflow still work in practice when they are run with fireworks or jobflow-remote. The data objects are a really important point
- [ ] should we provide a forcefield implementation for testing?
- [ ] should we provide a tutorial for pheasy? if we had a forcefield implementation, this would be much easier to do.
- [ ] conflicts need to be solved :)

esoteric-ephemera commented:
Hey @hrushikesh-s sorry I missed your messages. These schema changes are very much intentional:

> 1. The new emmet schema replaces the old PhononBSDOSDoc with a PhononBSDOSTask + built PhononBSDOSDoc split.

This is to explicitly allow for storing these as separate objects. `PhononBSDOSTask` should go to regular mongo collections, whereas band structures and DOS should go to GridFS

> 2. phonon_bandstructure and phonon_dos are no longer stored as raw pymatgen objects; they use emmet wrapper schemas (PhononBS, PhononDOS).

There is an [opt-in setting](https://github.com/materialsproject/emmet/blob/85b7edc3e1a8b4e49d7cc36ec4a7b7465bacc0dd/emmet-core/emmet/core/settings.py#L176) for using the new emmet models. The default is still to use the pymatgen models in atomate2 (opt-out).

> 3. The following legacy stored fields are not present as explicit stored fields in the new emmet schema:

    free_energies
    heat_capacities
    internal_energies
    entropies
    temperatures
    uuids
    jobdirs

There is no reason to store `free_energies`, `heat_capacities`, `internal_energies`, `entropies`, and `temperatures` in the document, since these are computed from a set of temperatures. What if the user wants a different set of temps than the one prescribed in the document?

The `uuids` and `jobdirs` are stored in a [`calc_meta` field](https://github.com/materialsproject/emmet/blob/85b7edc3e1a8b4e49d7cc36ec4a7b7465bacc0dd/emmet-core/emmet/core/phonon.py#L403)

> 5. has_imaginary_modes is now a computed field rather than a plain stored field.

A model computed field will still be saved in the document if it has been calculated. It is just a way to [defer calculating an expensive model field](https://docs.pydantic.dev/latest/concepts/fields/#the-computed_field-decorator) until the user requests it

Regarding the other fields being renamed, the document model explicitly handles the migration internally. You don't have to do anything to coerce the fields. Put a decent amount of work into it to do that since it needed to migrate our existing data seamlessly too

Please revert these changes regarding the schema or communicate what changes you need in the schema to get the workflows to run correctly

hrushikesh-s commented:
Hi @JaGeo and @esoteric-ephemera, thanks for your detailed comments. I will start working on the comments this Friday and will keep you updated.

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,953 @@
+    subprocess.call(shlex.split(pheasy_cmd_3))
+    subprocess.call(shlex.split(pheasy_cmd_4))
+
+    # When this code is run on Github tests, it is failing because it is
+    # not able to find the FORCE_CONSTANTS file. This is because the file is
+    # somehow getting generated in some temp directory. Can you fix the bug?
```
i guess this works now as well

JaGeo commented:
@hrushikesh-s do you have a suggestion for the timeline? More and more conflicts start to show up and it will only get harder to finally merge

JaGeo commented on tests/forcefields/flows/test_phonon.py:
```diff
@@ -216,47 +217,56 @@ def test_phonon_wf_force_field(
+        ],
         "free_energy": (
-            [11500.0, 11000.0, 9100.0, 6100.0, 2200.0]
+            [3164.0, 3053.0, 2351.0, 999.0, -868.0]
```
This changed so much as Chgnet was updated, right?

hrushikesh-s replied:
Yes, right. I think CHGNet changed from the old MPtrj-trained model to the new MatPES-trained model that matgl now serves.

JaGeo commented on src/atomate2/common/jobs/gruneisen.py:
```diff
@@ -50,7 +53,7 @@ def shrink_expand_structure(structure: Structure, perc_vol: float) -> Response:
     return Response(output={"plus": plus_struct, "minus": minus_struct})
 
 
-@job(data=[PhononBSDOSDoc])
```
Are you sure this should be removed?

JaGeo replied:
@hrushikesh-s i believe this is the very last open point

JaGeo commented on src/atomate2/common/jobs/pheasy.py:
```diff
@@ -0,0 +1,916 @@
+    ALM = None
+
+# CODATA 2018: 1 Angstrom = 1 / 0.529177210903 Bohr
+ANGSTROM_TO_BOHR = 1.8897261246257702
```
Are you using this at multiple places? If you, please make a module for it. Eg "utils.units.py"

JaGeo commented:
@hrushikesh-s can you please resolve the very last open point? I am happy to merge if this is done
